raiden/network/transport/matrix/transport.py

```diff
@@ -587,7 +587,11 @@ def _send_global(room_name, serialized_message):
                     JSONSerializer.serialize(message) for message in messages_for_room
                 )
                 _send_global(room_name, message_text)
-                self._global_send_queue.task_done()
+                for _ in messages_for_room:
+                    # Every message needs to be marked as done.
+                    # Unfortunately there's no way to do that in one call :(
+                    # https://github.com/gevent/gevent/issues/1436
+                    self._global_send_queue.task_done()
```

Comment:
This is not synchronized with the calls to `get`, which can get reordered above while grouping the messages per room, if an exception is raised the wrong tasks can be set as `done`.

It looks like `_global_send_queue` being a `JoinableQueue` is only useful for startup of the transport, because the global messages have to be sent first. Perhaps this can be simplified further ?

Reply:
There is no ordering in the JoinableQueue anyway. It's just a counter internally.